packages/options/src/option.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * An implementation of the Rust Option type in JavaScript.
+ * It can be one of the following:
+ * - <code>{@link Some}<T></code>: Meaning there is a value of type T.
+ * - <code>{@link None}</code>: Meaning there is no value.
+ */
+export type Option<T> = Some<T> | None;
+
+/**
+ * Defines a looser type that can be used when serializing an {@link Option}.
+ * This allows us to pass null or the Option value directly whilst still
+ * supporting the Option type for use-cases that need more type safety.
+ */
+export type OptionOrNullable<T> = Option<T> | T | null;
+
+/**
+ * Represents an option of type `T` that has a value.
+ *
+ * @see {@link Option}
+ */
+export type Some<T> = { __option: 'Some'; value: T };
+
+/**
+ * Represents an option of type `T` that has no value.
+ *
+ * @see {@link Option}
+ */
+export type None = { __option: 'None' };
```

Comment:
These _really, really_ need to be `Readonly`.
```suggestion
/**
 * Represents an option of type `T` that has a value.
 *
 * @see {@link Option}
 */
export type Some<T> = Readonly<{ __option: 'Some'; value: T }>;

/**
 * Represents an option of type `T` that has no value.
 *
 * @see {@link Option}
 */
export type None = Readonly<{ __option: 'None' }>;
```